### ISSUE TITLE
[$250] Editing comment with image attachment fails to load the image when cache expires

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#42206]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**
**Reproducible in staging?:**
**Reproducible in production?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @kidroca 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1715767781452439

## Action Performed:

1. Type text and add an image attachment.
2. After sending the comment, inspect the <img> tag for data-expensify attributes.
3. Edit the comment and make a trivial change to the text before the attachment.
4. Observe that the <img> tag no longer has data-expensify attributes.
5. Disable cache in dev tools, or clear cache and reload the page.
6. Observe that the edited comment no longer loads the image.
LINK: https://staging.new.expensify.com/r/5708510475809890

## Expected Result:
Editing a comment, clearing cache, and reloading the page should display the comment and the image attachment.

## Actual Result:
Clearing cache and reloading shows the comment and a placeholder box instead of the image attachment.
More details here: https://github.com/Expensify/App/issues/41952#issuecomment-2112095398

## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/d87424e2-9423-421c-8712-a71e4437c250

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~018082d3734c0fe3cf</li>
        <li>Upwork Job ID: 1792320613124546560</li>
        <li>Last Price Increase: 2024-06-10</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

**Comment by m-natarajan** [comment]
@mallen tagging you as per this [comment](https://expensify.slack.com/archives/C01GTK53T8Q/p1715795872566669?thread_ts=1715767781.452439&cid=C01GTK53T8Q)

**Comment by jliexpensify** [comment]
@m-natarajan you've tagged the wrong person. cc @mallenexpensify

**Comment by mallenexpensify** [comment]
Labeled internally and added to #vip-vsb cuz it doesn't have to do with money. All yours JLi

**Comment by jliexpensify** [comment]
Not overdue, waiting on an Engineer to pick this up.

**Comment by jliexpensify** [comment]
Not overdue

**Comment by rushatgabhane** [comment]
internal. not overdue

**Comment by rushatgabhane** [comment]
🎀 👀 🎀 assign internal

**Comment by amyevans** [comment]
Sorry, I don't have bandwidth to take this, unassigning!

**Comment by rushatgabhane** [comment]
🎀 👀 🎀 assign internal

**Comment by mountiny** [comment]
I am going to make this external and look for proposals, not sure if this really cannot be solved externally

**Comment by truph01** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Clearing cache and reloading shows the comment and a placeholder box instead of the image attachment.
More details here: #41952 (comment)

### What is the root cause of that problem?
When we edit the comment with image attachment, the `data-expensify-` attributes were stripped. The `data-expensify-source` is one of them, so after reloading the authentication header from `data-expensify-source` no longer applies, therefore the image fails to load.

### What changes do you think we should make in order to solve the problem?

Store the `data-expensify-` attributes in an object like a cache (or even in Onyx), then after editing, apply the attributes again to the image html data. This way we'll still preserve the important metadata of `data-expensify-` in the image tags.

### What alternative solutions did you explore? (Optional)
Prevent the user from editing text with attachment.

**Comment by mountiny** [comment]
@rushatgabhane what do you think of the proposal?

**Comment by rushatgabhane** [comment]
> When we edit the comment with image attachment, the data-expensify- attributes were stripped.

@truph01 your suggestion is to store the attributes and apply them again before saving the comment.


But why does editing a comment strip the expensify attributes in the first place? Can we make it so that they don't get removed? Then we won't have this problem

**Comment by truph01** [comment]
@rushatgabhane This is what the markdown for "text with image" looks like
```
an image

![Image name](https://www.expensify.com/chat-attachments/3998734689149153575/w_b90982bdb232e02e7629cb364264ac230a2f012a.jpg.1024.jpg)
```

As you can see, the markdown for image only contains the URL of the image and the image name (if any).

So editing the comment will strip the Expensify attributes because the Expensify attributes are not parts of the markdown syntax for images.

> Can we make it so that they don't get removed?

This will mean we need to use another markdown syntax for images that contain the attributes
Like 
```
![Image name data-expensify-source="https://www.expensify.com/chat-attachments/3998734689149153575/w_b90982bdb232e02e7629cb364264ac230a2f012a.jpg.1024.jpg" data-expensify-height="100" data-expensify-width="100"](https://www.expensify.com/chat-attachments/3998734689149153575/w_b90982bdb232e02e7629cb364264ac230a2f012a.jpg.1024.jpg)
```
Or another idea is to make the attributes the params of the image URL, like `?data-expensify-source=...&data-expensify-height=100`

Either way, IMHO that will make the image markdown very lengthy and hard to edit to the user, and might not conform to any markdown standard

**Comment by kidroca** [comment]
These attributes are added by the backend when an attachment is uploaded; they are not part of the original syntax. When the backend processes the attachment, it adds the necessary attributes.

I believe the fix should be on the backend. When editing a comment containing an attachment, the backend should apply the same logic and replace/insert the `<img>` HTML.

There's a related ticket addressing a similar problem with `<video>` attachments. @rushatgabhane, perhaps you can cross-reference how the team handled that issue and see if the same solution can be applied here.
- https://github.com/Expensify/App/issues/41952

**Comment by mountiny** [comment]
Working on the backend solution

**Comment by mountiny** [comment]
@rushatgabhane I am not sure if I am doing something wrong but I cannot reproduce this

**Comment by rushatgabhane** [comment]
Same! I cannot reproduce this bug

https://github.com/Expensify/App/assets/29673073/c65f05c3-2b58-447b-ad31-97c62b6d34c5

**Comment by rushatgabhane** [comment]
@mountiny i can consistently reproduce this bug on staging, but not on production.

### Staging - Reproducible

https://github.com/Expensify/App/assets/29673073/4f70fb63-b8f9-45f9-9df2-b2726f9edc11



### Production - Not Reproducible

https://github.com/Expensify/App/assets/29673073/b08a2d43-b22c-4865-839d-0517f1984ebf

**Comment by kidroca** [comment]
I can still recreate this bug on prod
Here's a short debug session where we can see the `<img>` in the updated comment lacks the `data-expensify` attribute. This is the reason we're getting the 302 redirects to an error page on old dot, because the request is meant to be made with an auth header, but isn't

https://github.com/Expensify/App/assets/12156624/d0b8271e-719e-440e-b31a-0c1d27b463ce

**Comment by rushatgabhane** [comment]
@kidroca that makes sense, after the auth for oldDot has expired the image won't load anymore.

**Comment by rushatgabhane** [comment]
@mountiny this is an internal bug

**Comment by rushatgabhane** [comment]
To reproduce this wait for auth token to expire on oldDot

**Comment by kidroca** [comment]
I don't think auth token matters - it's not begin sent regardless of whether it's expired or not

The network requests are redirected to Old Dot, because of some internal server error handling 
But the main issue is, the `<img>` html is getting re-written and some attributes are stripped upon editing the comment

**Comment by mountiny** [comment]
Thanks, I will have to investigate more

**Comment by mountiny** [comment]
No update right now, this is lower priority than other tasks

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by mountiny** [comment]
lets keep this open until another retest but seems like this might have gotten fixed elsewhere

**Comment by mallenexpensify** [comment]
I was able to reproduce on the latest staging build, Desktop, Version 9.0.25-10 
<img width="534" alt="image" src="https://github.com/user-attachments/assets/05a9c489-2c18-474a-9bf0-86f22bde60ea">

1. User A types message in DM with user B > adds image > clicks `Send`
2. User A edits text in message
3. User B views message, observes image doesn't show. 

<img width="534" alt="image" src="https://github.com/user-attachments/assets/5379f79f-30a4-47fe-9504-65d16cf4cbe3">

**Comment by mountiny** [comment]
I am not able to prioritize this at the beginning of this week unfortunately.

**Comment by mountiny** [comment]
Adding hot picks for this one as I am focusing on a fire fix in the foreseeable future and not able to prioritise this one right now

**Comment by mountiny** [comment]
same

**Comment by mountiny** [comment]
focusing on workspace feeds wave-collect

**Comment by jliexpensify** [comment]
Not overdue

**Comment by mountiny** [comment]
I was talking with @mjasikowski that he might be able to take this on later in the week

**Comment by jliexpensify** [comment]
Not overdue

**Comment by mountiny** [comment]
@mjasikowski will help out here 💪

**Comment by jliexpensify** [comment]
Assigned!

**Comment by mjasikowski** [comment]
FInally have time to dig into this.

**Comment by mountiny** [comment]
was able to repro still

https://github.com/user-attachments/assets/79bfbfcc-7303-480b-836d-f09effc54ccb

**Comment by mjasikowski** [comment]
Current findings:
- Web-Expensify does strip extra attributes form img tags via `sanitizeAndValidateComment`
- Unfortunately, Expensify-Common does this as well in `ExpensiMark.ts`, making the tags stripped before we even get to the `UpdateComment` API call

**Comment by mjasikowski** [comment]
There's a draft expensify-common PR now adding attribute caching support to `ExpensiMark.ts`, I'll try to do a full test  tomorrow.

**Comment by mjasikowski** [comment]
It turns out that only expensify-common fix is enough, and that's ready in https://github.com/Expensify/expensify-common/pull/808

**Comment by mjasikowski** [comment]
PR still under review

**Comment by mjasikowski** [comment]
https://github.com/Expensify/Web-Expensify/pull/43698 was needed after all, submitted that too

**Comment by jliexpensify** [comment]
Hi @mjasikowski - just a heads up to let you know I'll be OOO from the 3rd to 14th. 

I'll prep a Payment Summary, I think this is all that's needed from me:

- C+: @rushatgabhane $250 (New.Dot payment)

**Comment by mjasikowski** [comment]
@m-natarajan the fix is now on both staging and production, can you re-test?

**Comment by mallenexpensify** [comment]
I just did a quick-test in the latest staging build on Desktop and, after editing a comment, the image still shows!  Think we need to to still leave this open for a retest? 
<img width="476" alt="image" src="https://github.com/user-attachments/assets/e115c73a-ce3a-426d-baef-4b5b57ecf766">

**Comment by jliexpensify** [comment]
Payment Summary

- C+: @rushatgabhane $3250 (owed via ND payments)

**Comment by jliexpensify** [comment]
Ha, yeah I fat fingered that one - thanks @rushatgabhane!

**Comment by garrettmknight** [comment]
$250 approved for @rushatgabhane

**Comment by mallenexpensify** [comment]
@truph01 can you provide more details plz? Ideally a bullet point list with links and time _stamps_?  It'll make it easier to review for potential compensation.  Thx

**Comment by truph01** [comment]
@mallenexpensify Sure, here's the timeline:

- The issue was [made external](https://github.com/Expensify/App/issues/42206#issuecomment-2157821329) on June 10
- I [posted my proposal](https://github.com/Expensify/App/issues/42206#issuecomment-2157860875) right after that on June 10, which has the `Solution 2` of 
> Store the data-expensify- attributes in an object like a cache, with the image source as key, then after editing, apply the attributes again to the image html data. This way we'll still preserve the important metadata of data-expensify- in the image tags.

And some draft implementation details

- On June 14, There was [a suggestion to fix in back-end](https://github.com/Expensify/App/issues/42206#issuecomment-2167681024) by @kidroca
- @mountiny [seems to agree with it](https://github.com/Expensify/App/issues/42206#issuecomment-2168364096) and the issue is considered to be `Internal` and back-end fix was attempted from this point.

- On Sep 23, @mountiny [assigned](https://github.com/Expensify/App/issues/42206#issuecomment-2367440420) @mjasikowski to work on the issue

- On Sep 26, [the PR](https://github.com/Expensify/expensify-common/pull/808) was raised by @mjasikowski, which uses same approach as [my proposal](https://github.com/Expensify/App/issues/42206#issuecomment-2157860875). No back-end fix is done.

> This PR adds caching for additional attributes to the <img> tags when converted to and back from Markdown. 

As such, I think a compensation of 50% to me makes sense, because the issue was made Internal for attempted back-end fix but actually there's no back-end fix done and the solution that ends up being used is same as suggested in my proposal.

**Comment by mountiny** [comment]
> No back-end fix is done.

backend fix was done, you just do not have access to it https://github.com/Expensify/Web-Expensify/pull/43729

**Comment by truph01** [comment]
@mountiny Was it needed to fix this issue? As from my testing with my initial proposal, only the front-end fix is enough.

**Comment by mountiny** [comment]
It was required as we were stripping the necessary tags in PHP layer too so not only the FE changes would be sufficient for all cases. @mjasikowski will be able to provide better explanation

**Comment by mjasikowski** [comment]
The tags were being stripped in the PHP layer. Caching attributes was already implemented in App for videos and that needed to be adjusted to apply images as well, but the was only a small portion of the changes needed.

**Comment by truph01** [comment]
Thanks for your input, I've updated the [timeline summary](https://github.com/Expensify/App/issues/42206#issuecomment-2447911021) highlighting there's a back-end fix.

Looking forward to the final decision from @mountiny @mallenexpensify 

Thanks in advance!

**Comment by mountiny** [comment]
Sorry @truph01, I don't think we are going to issue a reward for this one as there were many pieces required to fix this, and it was handled by @mjasikowski regardless of your proposal. 

We appreciate your contributions and hope you will continue fixing issues in our repo ahead.
